pennylane/templates/subroutines/semi_adder.py

```diff
@@ -0,0 +1,285 @@
+        return {
+            "num_y_wires": len(self.hyperparameters["y_wires"]),
+        }
+
+    @property
+    def num_params(self):
+        return 0
+
+    def _flatten(self):
+        metadata = tuple((key, value) for key, value in self.hyperparameters.items())
+        return tuple(), metadata
+
+    @classmethod
+    def _unflatten(cls, data, metadata):
+        hyperparams_dict = dict(metadata)
+        return cls(**hyperparams_dict)
+
+    def map_wires(self, wire_map: dict):
+        new_dict = {
+            key: [wire_map.get(w, w) for w in self.hyperparameters[key]]
+            for key in ["x_wires", "y_wires", "work_wires"]
+        }
+
+        return SemiAdder(
+            new_dict["x_wires"],
+            new_dict["y_wires"],
+            new_dict["work_wires"],
+        )
+
+    def decomposition(self):  # pylint: disable=arguments-differ
+        r"""Representation of the operator as a product of other operators."""
+        return self.compute_decomposition(**self.hyperparameters)
+
+    @classmethod
+    def _primitive_bind_call(cls, *args, **kwargs):
+        return cls._primitive.bind(*args, **kwargs)
+
+    @staticmethod
+    def compute_decomposition(x_wires, y_wires, work_wires):  # pylint: disable=arguments-differ
+        r"""Representation of the operator as a product of other operators.
+        The implementation is based on `arXiv:1709.06648 <https://arxiv.org/abs/1709.06648>`_.
+
+        Args:
+            x_wires (Sequence[int]): the wires that store the integer :math:`x`
+            y_wires (Sequence[int]): the wires that store the integer :math:`y`
+            work_wires (Sequence[int]): the auxiliary wires to use for the addition. Exactly ``len(y_wires) - 1`` work
+                wires should be provided.
+
+        Returns:
+            list[.Operator]: Decomposition of the operator
+        """
+        op_list = []
+
+        # revert wires to follow PennyLane convention
+
+        x_wires_pl = x_wires[::-1][: len(y_wires)]
+        y_wires_pl = y_wires[::-1]
+        work_wires_pl = work_wires[::-1]
+        op_list.append(qml.TemporaryAND([x_wires_pl[0], y_wires_pl[0], work_wires_pl[0]]))
+
+        for i in range(1, len(y_wires_pl) - 1):
+            if i < len(x_wires_pl):
+                op_list += _left_operator(
+                    work_wires_pl[i - 1], x_wires_pl[i], y_wires_pl[i], work_wires_pl[i]
+                )
+            else:
+                # If the number of qubits in |x> is smaller than |y>, we can conceptually complete the |x> state
+                # with |0> on the left making sure they are of the same size. Assuming this, we can simplify the left operator.
+                op_list.append(qml.CNOT([work_wires_pl[i - 1], y_wires_pl[i]]))
+                op_list.append(
+                    qml.TemporaryAND([work_wires_pl[i - 1], y_wires_pl[i], work_wires_pl[i]])
+                )
+                op_list.append(qml.CNOT([work_wires_pl[i - 1], work_wires_pl[i]]))
```

Comment:
I think it might actually be cleaner if this is moved into the `_left_operator` function. You can add a flag like `ik_is_zero=False` by default.

Reply:
nice suggestion 🙌